--- cryptsetup-1.2.0.old/lib/libdevmapper.c	2012-02-29 13:01:29.000000000 +0100
+++ cryptsetup-1.2.0/lib/libdevmapper.c	2012-02-29 13:05:48.645106899 +0100
@@ -172,6 +172,13 @@ static char *__lookup_dev(char *path, de
 		    !strncmp(entry->d_name, "..", 2))
 			continue;
 
+		if (dir_level == 0 &&
+                    (!strcmp(entry->d_name, "shm") ||
+                     !strcmp(entry->d_name, "fd") ||
+                     !strcmp(entry->d_name, "char") ||
+                     !strcmp(entry->d_name, "pts")))
+                        continue;
+
 		strncpy(ptr, entry->d_name, space);
 		if (stat(path, &st) < 0)
 			continue;
